lib/Zonemaster/Backend/ErrorMessages.pm

```diff
@@ -0,0 +1,88 @@
+
+package Zonemaster::Backend::ErrorMessages;
+
+
+use Readonly;
+use Locale::TextDomain qw[Zonemaster-Backend];
+
+Readonly my @CUSTOM_MESSAGES_CONFIG => (
+    {
+        pattern => "/(domain|hostname)",
+        config => {
+            string => {
+                pattern => N__ 'The domain name character(s) are not supported'
+            }
+        }
+    },
+    {
+        pattern => "/nameservers/\\d+/ip",
+        config => {
+            string => {
+                pattern => N__ 'Invalid IP address'
+            }
+        }
+    },
+    {
+        pattern => "/nameservers/\\d+/ns",
+        config => {
+            string => {
+                pattern => N__ 'The domain name character(s) are not supported'
+            }
+        }
+    },
+    {
+        pattern => "/ds_info/\\d+/keytag",
+        config => {
+            integer => {
+                type => N__ 'Keytag should be a positive integer',
+                minimum => N__ 'Keytag should be a positive integer'
+            }
+        }
+    },
+    {
+        pattern => "/ds_info/\\d+/algorithm",
+        config => {
+            integer => {
+                type => N__ 'Algorithm should be a positive integer',
+                minimum => N__ 'Algorithm should be a positive integer'
+            }
+        }
+    },
+    {
+        pattern => "/ds_info/\\d+/digtype",
+        config => {
+            integer => {
+                type => N__ 'Digest type should be a positive integer',
+                minimum => N__ 'Digest type should be a positive integer'
+            }
+        }
+    },
+    {
+        pattern => "/ds_info/\\d+/digest",
+        config => {
+            string => {
+                pattern => N__ 'Invalid digest format'
+            }
+        }
+    },
+    {
+        pattern => "/language",
+        config => {
+            string => {
+                pattern => N__ 'Invalid language tag format'
+            }
+        }
+    },
+    {
+        pattern => ".*",
```

Comment:
Forgive me for jumping the gun here. I've haven't read through all changes and thought about all the consequences yet, so I don't have a clear view of what I'm looking at. But I'm not getting a good feeling about this general pattern matching feature. Especially with this catchall. It seems really easy to forget making sure that the parameters are appropriately covered. At the same time it seems less straight forward to actually check whether or not they are.

But maybe (hopefully) I'm wrong and this is actually easier to maintain than it looks to me?

Reply:
> But I'm not getting a good feeling about this general pattern matching feature.

I don't really know what else would be a good way to add custom messages support then. The pattern matching is used to provide custom messages on a per-path basis, the inner structure resembles what the JSON::Validator uses to handle error messages. I will provide documentation for the structure.

Otherwise I am open to suggestion if someone see an other solution to implement the feature. What I also thought of what to directly modify the validation module to add something that stores the error message in the JSON schema (by extending [the json schema](https://json-schema.org/understanding-json-schema/reference/schema.html#id3) and the  [schema class](https://metacpan.org/pod/JSON::Validator::Schema#Make-a-new-validation-class) and the validation class). That would work for specific error messages but not generic ones like `Missing properties` or `Wrong property type`, which may ideally require to embed some of the translation logic directly into the module. But that would definitely better, I totally agree that the pattern matching to customize the error message is more a hack or a work around that anything else.

> Especially with this catchall.

This is to handle generic common error messages that can be present for any element in the json object (hence matching all path).

> But maybe (hopefully) I'm wrong and this is actually easier to maintain than it looks to me.

I don't think it is very difficult to maintain, the main drawback would be that the error message are defined in a place away from the schema definition, which can make it hard to track.